xarray/tests/test_concat.py

```diff
@@ -1261,3 +1262,103 @@ def test_concat_index_not_same_dim() -> None:
         match=r"Cannot concatenate along dimension 'x' indexes with dimensions.*",
     ):
         concat([ds1, ds2], dim="x")
+
+
+def test_concat_join_coordinate_variables_non_asked_dims():
+    ds1 = Dataset(
+        coords={
+            "x_center": ("x_center", [1, 2, 3]),
+            "x_outer": ("x_outer", [0.5, 1.5, 2.5, 3.5]),
+        },
+    )
+
+    ds2 = Dataset(
+        coords={
+            "x_center": ("x_center", [4, 5, 6]),
+            "x_outer": ("x_outer", [4.5, 5.5, 6.5]),
+        },
+    )
+
+    # Using join='outer'
+    expected_wrongly_concatenated_xds = Dataset(
+        coords={
+            "x_center": ("x_center", [1, 2, 3, 4, 5, 6]),
+            "x_outer": ("x_outer", [0.5, 1.5, 2.5, 3.5, 4.5, 5.5, 6.5]),
+        },
+    )
+    # Not using strict mode will allow the concatenation to surprisingly happen
+    # even if `x_outer` sizes do not match
+    actual_xds = concat(
+        [ds1, ds2],
+        dim="x_center",
+        data_vars="different",
+        coords="different",
+        join="outer",
+    )
+    assert all(actual_xds == expected_wrongly_concatenated_xds)
+
+    # Using join='strict'
+    # A check similar to the one made on non-indexed dimensions regarding their sizes.
```

Comment:
Great explanation, thanks.

What's the difference between `strict` and `exact`? Could we show the difference?

---

Assuming we do merge this, we'll want this sort of explanation in the docs...

Reply:
> What's the difference between strict and exact? Could we show the difference?

**TLDR: join='exact` does not consider dimension names when aligning. join='strict' expects matching dimension names when aligning.**

The best difference is exposed in [`test_align_exact_vs_strict`](https://github.com/pydata/xarray/pull/8698/files#diff-ca5c2de2fe6e9e25fbf22bd53e4976c15da74900dfb14deb7e6e87f5377230e3R3197) . We can see an instance of two DataArrays  whom alignment is accepted with join='exact', but not with join='strict'. 'exacts' allows _shape-compatible_ DataArrays to be aligned with differing dimension names, whereas 'strict' forbids.

To be fair, I would have expected that this more restrictive logic would just have been added to the existing 'exact' mode. But when I tried so, many tests were failing. Also, I only tested on toy examples. I was quite unfamiliar with this alignment logic so I sticked to following the desired behaviour described in https://github.com/pydata/xarray/issues/6806 ; and trusted the existing test coverage for the rest.

I would say the biggest addition is this new check that is performed early during the aligment operation:

https://github.com/pydata/xarray/pull/8698/files#diff-c32c8ede32701deaba092f13d721242d5ef69144b2d5e76ecf1748333f1106f5R595

It is only applied for the `strict` mode, and checks that all objects to be aligned have the exact same dimensions. It is really a "symbolic" check on the dimension names, to ensure that compatible arrays with same shape but different dimension names will not _slip through the cracks_. It is pretty restrictive and add more constraints, so that's probably be why this mode would be an opt-in as it could annoy some users.

Reply:
Thanks a lot. I added another question to that example.